### PR TITLE
rpc server with HTTP/WS on the same socket

### PR DESCRIPTION
**Pull request by niklasad1**
This PR adds a RPC server with similar functionality as the current WS server but it supports both HTTP and WS on the same socket with some new functionality such as tower middleware.

The long-term goal is to remove the HTTP RPC server completely and rely on the "mixed WS/HTTP" server but the HTTP server is kept for backward compatibility giving folks some time to migrate tools and similar to the `mixed WS/HTTP server`.

This release has plenty of changes and can be found [here](https://github.com/paritytech/jsonrpsee/blob/v0.16.0/CHANGELOG.md#v0160---2022-11-09)

polkadot companion: https://github.com/paritytech/polkadot/pull/6339
cumulus companion: https://github.com/paritytech/cumulus/pull/1905

### CODE REVIEW

File: client/service/src/lib.rs [hunk 1]
```diff
@@ -346,45 +333,52 @@ where
...
-		ws_config,
+		server_config.clone(),
+		metrics.clone(),
+		gen_rpc_module(deny_unsafe(ws_addr, &config.rpc_methods))?,
+		config.tokio_handle.clone(),
+		rpc_id_provider,
+		"WS",
+	);
+
+	let http_fut = sc_rpc_server::start_server(
```
Comment (niklasad1) [inline client/service/src/lib.rs:362]: NOTE: for backwards compatibility just to start another server on port 9333 i.e, `--rpc-port` I would want to kill remove it completely but that would break lots of things :P

Comment (niklasad1) [reply client/service/src/lib.rs:362]: I could also use the old http server i.e, jsonrpsee v0.15 but I would prefer not to.

File: client/rpc-servers/src/lib.rs [hunk 2]
```diff
@@ -77,87 +80,49 @@ impl WsConfig {
...
+		} else {
+			// allow all cors
+			CorsLayer::permissive()
+		}
 	};
 
-	let mut builder = WsServerBuilder::new()
+	let middleware = tower::ServiceBuilder::new()
+		// Proxy `GET /health` requests to internal `system_health` method.
+		.layer(ProxyGetRequestLayer::new("/health", "system_health")?)
```
Comment (lexnv) [inline client/rpc-servers/src/lib.rs:114]: DQ: previously we exposed the `/health` API path for HTTP only. If we upgrade the HTTP connection to WS, would the `/health` API still be reachable for the "old-HTTP port, now WS port"?

File: client/service/src/lib.rs [hunk 3]
```diff
@@ -326,8 +314,7 @@ fn start_rpc_servers<R>(
 where
 	R: Fn(sc_rpc::DenyUnsafe) -> Result<RpcModule<()>, Error>,
 {
-	let (max_request_size, ws_max_response_size, http_max_response_size) =
-		legacy_cli_parsing(config);
+	let (max_request_size, ws_max_response_size, _) = legacy_cli_parsing(config);
```
Comment (lexnv) [inline client/service/src/lib.rs:317]: Nit: Would it make sense to extract `http_max_response_size` here and then create two separate `sc_rpc_server::Config`?

It seems that both servers do receive a `ws_max_response_size` now

Comment (niklasad1) [reply client/service/src/lib.rs:317]: nice catch, yeah you are right

File: client/rpc-servers/src/lib.rs (hunk 2 above)
Comment (niklasad1) [reply client/rpc-servers/src/lib.rs:114]: `/health` is enabled for both ports but as we discussed offline

File: client/rpc-servers/src/lib.rs [hunk 4]
```diff
@@ -86,31 +89,28 @@ pub async fn start_http<M: Send + Sync + 'static>(
...
+		.layer(cors.clone());
+
+	let builder = ServerBuilder::new()
+		.max_request_body_size(max_payload_in)
+		.max_response_body_size(max_payload_out)
+		.set_host_filtering(host_filter)
+		.set_middleware(middleware)
+		.custom_tokio_runtime(rt)
+		.http_only();
```
Comment (niklasad1) [inline client/rpc-servers/src/lib.rs:109]: this runs only on HTTP and will be deprecated/removed in few releases.

/cc @PierreBesson @jam206  would be great if you can migrate to `health checks` to the `ws-port` when this is deployed that should work

I think the most reasonable is to keep `--rpc-port` and similar flags and just deprecate the `ws-related` flags.
Then make the `9944` the default port.

File: client/rpc-servers/src/lib.rs [hunk 5]
```diff
@@ -179,23 +177,14 @@ pub async fn start_ws<M: Send + Sync + 'static>(
 	};
 
 	log::info!(
-		"Running JSON-RPC WS server: addr={}, allowed origins={:?}",
+		"Running JSON-RPC WS server: addr={}, cors={:?}",
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply client/rpc-servers/src/lib.rs:180]: also I kept `Running JSON-RPC WS server: addr={}` because we have plenty of scripts that parses this to know when the RPC servers has been started.

File: client/rpc-servers/src/lib.rs [hunk 6]
```diff
@@ -120,44 +120,44 @@ pub async fn start_http<M: Send + Sync + 'static>(
...
 		cors
 	);
 
 	Ok(handle)
 }
 
-/// Start WS server listening on given address.
-pub async fn start_ws<M: Send + Sync + 'static>(
+/// Start a JSON-RPC server listening on given address that supports both HTTP and WS.
+pub async fn start<M: Send + Sync + 'static>(
```
Comment (niklasad1) [inline client/rpc-servers/src/lib.rs:132]: we can enable folks to inject their own tower middleware now which could useful I reckon but we could do it in another PR

**Review by lexnv — APPROVED** [review]
LGTM! 👍  It also needs companions in Polkadot and Cumulus.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2084451

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2099992

**Review by sam0x17 — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

File: utils/frame/remote-externalities/src/lib.rs [hunk 7]
```diff
@@ -418,17 +443,32 @@ where
...
+						log::warn!(
+							target: LOG_TARGET,
+							"key {:?} had none corresponding value.",
+							&key
+						);
+						child_kv_inner.push((key.clone(), StorageData(vec![])));
+					},
+					Err(e) => {
+						log::error!(target: LOG_TARGET, "key {:?} failed: {:?}", &key, e);
+						batch_success = false;
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs:461]: //cc @kianenigma do we want to log "all failed calls in the batch" or just the first error? 

I did that to log all failed calls in the batch

[the remark this answers is not in the document]
Comment (niklasad1) [reply utils/frame/remote-externalities/src/lib.rs:461]: I will take care of that in @kianenigma's pending try-runtime PR, ignoring for now.

**Comment by niklasad1** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-36/1529/1

**Comment by PierreBesson** [comment]
I think it would be valuable to add a simple migration guide. As a node operator, it is not very easy to understand how to adapt configuration to this change and which flags will be deprecated. Also what should be the new canonical RPC port: 9933 (former http-port) or 9944 (former ws-port) ?
